### PR TITLE
Semver 2.0 

### LINKED ISSUE #557: Exclude prereleases from SemVer range maximums
Per e.g. https://github.com/npm/node-semver?tab=readme-ov-file#x-ranges-12x-1x-12- / https://github.com/npm/node-semver/commit/059a5adec5aefaa764bf4fdb717ae8b42cbbefcd - we need to exclude prereleases from range maximums as a rule. The simple way to do this is representing a range maximum with a prerelease of 0 (e.g., 1.x becomes >=1.0.0 <2.0.0-0).

See discussion in https://github.com/intersystems-community/zpm-registry/issues/91

I'd recommend a test-driven development workflow for this, expanding on the existing SemVer unit tests. This is v0.9.0 critical.

### PR DESCRIPTION
**Pull request by isc-shuliu**
Fix #557

### CODE REVIEW

**Review by isc-tleavitt — APPROVED** [review]

File: tests/unit_tests/Test/PM/Unit/SemVer/Expressions.cls
Comment (isc-kiyer) [inline tests/unit_tests/Test/PM/Unit/SemVer/Expressions.cls:37]: @isc-tleavitt @isc-shuliu  This may cause us some problems with maintenance releases where we currently rely on this behavior. If we don't want to support this, our release tool will need to lock down versions in not just the base module but every dependency.
For example, HealthConnect 7.0.0 lists HS.REST 1.1.0-1.m1 and HealthConnect.Common 4.1.1-1.m1 as fixed version dependencies. However, HealthConnect.Common lists HS.REST ^1.0.0 as a dependency. This would now break unless HealthConnect.Common also lists a specific version as a dependency.
We need the ability to toggle this behavior on vs off I think for HS. Either that or come up with an alternate solution for maintenance releases.
cc @isc-eneil

Comment (isc-tleavitt) [reply tests/unit_tests/Test/PM/Unit/SemVer/Expressions.cls:37]: @isc-kiyer that's poor use of semantic versioning to say that a prerelease is sequentially after a main release. If there's an alternate solution for maintenance releases that would be preferable.

Comment (isc-tleavitt) [reply tests/unit_tests/Test/PM/Unit/SemVer/Expressions.cls:37]: Specifically, I think this comes down to discipline around saying what's really a bugfix vs. enhancement, and making decisions about releases of individual packages based on their contents. Under what circumstances would we not want a top-level release to include fixes to known bugs in a dependency under a given minor release?

Comment (isc-tleavitt) [reply tests/unit_tests/Test/PM/Unit/SemVer/Expressions.cls:37]: All that said, we could address this with a top-level command and configurable "include prerelease" switch similar to includePrerelease in node-semver.

Comment (isc-tleavitt) [reply tests/unit_tests/Test/PM/Unit/SemVer/Expressions.cls:37]: So question is, among the following options, where does an "include prerelease" switch belong?
* module.xml, as a subelement of the module (or could *perhaps* jump through hoops in XSLTs to support via an attribute of Dependencies) - use case: always make it explicit for top-level packages in our package-lock-ish use case for e.g. HealthConnect, HSBASE.
* option in "load" and "install" commands - use case: in a proper SemVer environment or possibly a CI workflow, pick up not-yet-released upcoming versions of dependencies for testing
* system-level setting - use case: development environments where we'll always want to be able to pick up prereleases

My gut feeling is the system level setting, because it'll cover all cases and I want to do as little work as possible to enable poor use of SemVer.

Comment (isc-kiyer) [reply tests/unit_tests/Test/PM/Unit/SemVer/Expressions.cls:37]: I think a system level setting is a good idea. That being said, I would like to perhaps revisit just taking the next patch version (even if including other patches along the way) with the HS team to see if we can change maintenance releases to do that instead. Copying @bourette

Comment (bourette) [reply tests/unit_tests/Test/PM/Unit/SemVer/Expressions.cls]: I raised this with Jaideep, perhaps we could do the same with I4H/HC.

File: src/cls/IPM/General/SemanticVersionExpression/Comparator.cls
Comment (isc-tleavitt) [inline src/cls/IPM/General/SemanticVersionExpression/Comparator.cls:96]: I would recommend coercing this to be a number with:
`If (+##class(%IPM.Repo.UniversalSettings).GetValue("SemVerPostRelease") = 0) || ($Extract(..Operator,1) = "=")  {`

That way null is 0 (the default)

File: tests/unit_tests/Test/PM/Unit/SemVer/Expressions.cls
Comment (bourette) [reply tests/unit_tests/Test/PM/Unit/SemVer/Expressions.cls]: Just for completeness, we moved from + to - when we realized that re-spinning a release caused it to pick up newer code, making it impossible to reproduce a kit.

File: tests/unit_tests/Test/PM/Unit/SemVer/Abstract.cls
Comment (isc-tleavitt) [inline tests/unit_tests/Test/PM/Unit/SemVer/Abstract.cls:1]: Nicely done.

**Review by isc-tleavitt — CHANGES_REQUESTED** [review]
One minor change requested (just in case the installer doesn't run).

**Comment by isc-shuliu** [comment]
Actually, b9552ee should be part of #560 instead of this one. Will cherry pick there and revert on this branch.

**Comment by isc-tleavitt** [comment]
> Actually, [b9552ee](https://github.com/intersystems/ipm/commit/b9552ee494415a3879236e9bcbed82ad58e3e0db) should be part of #560 instead of this one. Will cherry pick there and revert on this branch.

It really belongs on top of both of them; the two seem a bit tangled now. Will just wait until both are approved to merge.

**Review by isc-tleavitt — APPROVED** [review]

**Review by isc-kiyer — APPROVED** [review]

**Pull request merged**
